src/geometry/GeoTrackView.i.hh

```diff
@@ -74,54 +75,108 @@ GeoTrackView& GeoTrackView::operator=(const DetailedInitializer& init)
     return *this;
 }
 
 //---------------------------------------------------------------------------//
 //! Find the distance to the next geometric boundary.
-CELER_FUNCTION void GeoTrackView::find_next_step()
+CELER_FUNCTION void
+GeoTrackView::find_next_step(vecgeom::VPlacedVolume const* pplvol)
 {
-    const vecgeom::LogicalVolume* logical_vol
-        = this->volume().GetLogicalVolume();
-    CELER_ASSERT(logical_vol);
+    CELER_ASSERT(pplvol);
+
     const vecgeom::VNavigator* navigator
-        = this->volume().GetLogicalVolume()->GetNavigator();
+        = pplvol->GetLogicalVolume()->GetNavigator();
     CELER_ASSERT(navigator);
 
-    next_step_
-        = navigator->ComputeStepAndPropagatedState(detail::to_vector(pos_),
-                                                   detail::to_vector(dir_),
-                                                   vecgeom::kInfLength,
-                                                   vgstate_,
-                                                   vgnext_);
+    if (this->is_outside())
+    {
+        // handling points outside of world volume
+        const real_type large = vecgeom::kInfLength;
+        next_step_            = pplvol->DistanceToIn(
+            detail::to_vector(pos_), detail::to_vector(dir_), large);
+        vgnext_.Clear();
+        if (next_step_ < large)
+        {
+            vgnext_.Push(pplvol);
+        }
+    }
+    else
+    {
+        next_step_
+            = navigator->ComputeStepAndPropagatedState(detail::to_vector(pos_),
+                                                       detail::to_vector(dir_),
+                                                       vecgeom::kInfLength,
+                                                       vgstate_,
+                                                       vgnext_);
+    }
+    dirty_ = false;
 }
 
 //---------------------------------------------------------------------------//
 //! Move to the next boundary and update volume accordingly
-CELER_FUNCTION void GeoTrackView::move_next_step()
+CELER_FUNCTION real_type GeoTrackView::move_to_boundary()
 {
+    if (dirty_)
+        this->find_next_step();
+
     // Move the next step plus an extra fudge distance
+    real_type dist = next_step_ + this->extra_push();
+    axpy(dist, dir_, &pos_);
+    next_step_ = 0.;
+    this->move_next_volume();
+    return dist;
+}
+
+//---------------------------------------------------------------------------//
+//! Move to the next boundary and update volume accordingly
+CELER_FUNCTION real_type GeoTrackView::move_next_step()
+{
+    if (dirty_)
+        this->find_next_step();
+    real_type dist = next_step_;
     axpy(next_step_, dir_, &pos_);
+    next_step_ = 0.;
     this->move_next_volume();
+    return dist;
+}
+
+//---------------------------------------------------------------------------//
+//! Move to the next boundary and update volume accordingly
+CELER_FUNCTION real_type GeoTrackView::move_by(real_type dist)
```

Comment:
I think the documentation is incorrect, and the name doesn't reflect what the function does. Can you at least fix the docs? Since this is a function definition and not inside a class declaration, prefer the long-form doc block:
```c++
/*!
 *  Blah blah.
 *
 * Additional long documentation...
 */
```